Neos.Neos/Classes/Domain/Workspace/WorkspacePublisher.php

```diff
@@ -0,0 +1,283 @@
+        );
+
+        $this->publishNodes(
+            $contentRepository,
+            $command->workspaceName,
+            $nodeIdsToPublish
+        );
+
+        return count($nodeIdsToPublish);
+    }
+
+    /**
+     * @return int The amount of changes that were published
+     */
+    public function publishDocument(PublishDocument $command): int
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($command->contentRepositoryId);
+
+        $nodeIdsToPublish = $this->resolveNodeIdsToPublishOrDiscard(
+            $contentRepository,
+            $command->workspaceName,
+            $command->documentId,
+            NodeTypeNameFactory::forDocument()
+        );
+
+        $this->publishNodes(
+            $contentRepository,
+            $command->workspaceName,
+            $nodeIdsToPublish
+        );
+
+        return count($nodeIdsToPublish);
+    }
+
+    /**
+     * @return int The amount of changes that were discarded
+     */
+    public function discardSite(DiscardSite $command): int
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($command->contentRepositoryId);
+        $nodeIdsToDiscard = $this->resolveNodeIdsToPublishOrDiscard(
+            $contentRepository,
+            $command->workspaceName,
+            $command->siteId,
+            NodeTypeNameFactory::forSite()
+        );
+
+        $this->discardNodes(
+            $contentRepository,
+            $command->workspaceName,
+            $nodeIdsToDiscard,
+        );
+
+        return count($nodeIdsToDiscard);
+    }
+
+    /**
+     * @return int The amount of changes that were discarded
+     */
+    public function discardDocument(DiscardDocument $command): int
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($command->contentRepositoryId);
+        $nodeIdsToDiscard = $this->resolveNodeIdsToPublishOrDiscard(
+            $contentRepository,
+            $command->workspaceName,
+            $command->documentId,
+            NodeTypeNameFactory::forDocument()
+        );
+
+        $this->discardNodes(
+            $contentRepository,
+            $command->workspaceName,
+            $nodeIdsToDiscard
+        );
+
+        return count($nodeIdsToDiscard);
+    }
+
+    private function publishNodes(
+        ContentRepository $contentRepository,
+        WorkspaceName $workspaceName,
+        NodeIdsToPublishOrDiscard $nodeIdsToPublish
+    ): void {
+        /**
+         * TODO: only rebase if necessary!
+         * Also, isn't this already included in @see WorkspaceCommandHandler::handlePublishIndividualNodesFromWorkspace ?
+         */
+        $contentRepository->handle(
+            RebaseWorkspace::create(
+                $workspaceName
+            )
+        )->block();
+
+        $contentRepository->handle(
+            PublishIndividualNodesFromWorkspace::create(
+                $workspaceName,
+                $nodeIdsToPublish
+            )
+        )->block();
```

Comment:
async or not async this is the question. Returning a promise for each handler is what we probably want in the long run but out of scope for now.

Reply:
Sync.